### PR TITLE
8311604: Simplify NOCOOPS requested addresses for archived heap objects

### PR DESCRIPTION
**Pull request by iklam**
This PR attempts to clean up some of the cruds in the existing code:

- Simplified the calculation of "requested address" when `UseCompressedOops` is disabled -- the archived heap objects are always written starting from 0x10000000
- Removed `HeapShared::to_requested_address()` so we don't have two kinds of "requested address"
- Updated the comments about "source" vs "buffered" vs "requested" addresses in archiveHeapWriter.hpp
- Removed `SerializeClosure::oop()` as the only oop we need to store into the archive header is `HeapShared::roots()`, which can be handled more easily with `FileMapHeader::_heap_roots_offset`
- Removed some unnecessary dependencies on `G1CollectedHeap::heap()->reserved()`

Also:
- Moved SerializeClosure to its own header file to improve build time.
- Fixed DeterministicDump.java, which wasn't archiving Java objects when  `UseCompressedOops` was disabled.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311604](https://bugs.openjdk.org/browse/JDK-8311604): Simplify NOCOOPS requested addresses for archived heap objects (**Enhancement** - P4)

### Reviewers
 * @ashu-mehra (no known openjdk.org user name / role) ⚠️ Review applies to [88d2b338](https://git.openjdk.org/jdk/pull/14792/files/88d2b338723aac872961454af1b1915ab80f31fe)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [fb8a2165](https://git.openjdk.org/jdk/pull/14792/files/fb8a216539241d57c3bca943251abc41a98160e6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14792/head:pull/14792` \
`$ git checkout pull/14792`

Update a local copy of the PR: \
`$ git checkout pull/14792` \
`$ git pull https://git.openjdk.org/jdk.git pull/14792/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14792`

View PR using the GUI difftool: \
`$ git pr show -t 14792`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14792.diff">https://git.openjdk.org/jdk/pull/14792.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14792#issuecomment-1624443395)

### CODE REVIEW

File: src/hotspot/share/cds/archiveHeapWriter.hpp [hunk 1]
```diff
@@ -38,26 +38,81 @@
...
+  //  See archiveBuilder.hpp.)
+  //
+  // - "source objects" are regular Java objects allocated during the execution
+  //   of "java -Xshare:dump". They can be used as regular oops.
+  //
+  //   HeapShared::archive_objects() recursively searches for the oops that need to be
+  //   stored into the CDS archive. These are entered into HeapShared::archived_object_cache().
+  //
+  // - "buffered objects" are copies of the "source objects", and are stored in into
+  //   ArchiveHeapWriter::_buffer, which is a GrowableArray that sites outside of
```
Comment (ashu-mehra) [inline src/hotspot/share/cds/archiveHeapWriter.hpp:80]: s/sites/sits

File: src/hotspot/share/cds/filemap.cpp [hunk 2]
```diff
@@ -2086,7 +2069,7 @@ address FileMapInfo::heap_region_requested_address() {
   } else {
     // We can avoid relocation if each region is mapped into the exact same address
     // where it was at dump time.
-    return /*dumptime*/header()->heap_begin() + r->mapping_offset();
+    return (address)ArchiveHeapWriter::NOCOOPS_REQUESTED_BASE;
```
Comment (ashu-mehra) [inline src/hotspot/share/cds/filemap.cpp:2072]: Can you please update the comment above this line as well to remove `each region` as we only have single heap region now.

**Review by ashu-mehra — APPROVED** [review]

**Comment by ashu-mehra** [comment]
minor nitpicks, otherwise looks good!

File: src/hotspot/share/cds/filemap.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/filemap.cpp:2073]: Thanks for the review. I updated the comment to clarify that this address was hard-coded at dump time, and explained the reason for that.

File: src/hotspot/share/cds/filemap.cpp [hunk 3]
```diff
@@ -286,8 +275,6 @@ void FileMapHeader::print(outputStream* st) {
   st->print_cr("- compressed_class_ptrs:          %d", _compressed_class_ptrs);
   st->print_cr("- cloned_vtables_offset:          " SIZE_FORMAT_X, _cloned_vtables_offset);
   st->print_cr("- serialized_data_offset:         " SIZE_FORMAT_X, _serialized_data_offset);
-  st->print_cr("- heap_begin:                     " INTPTR_FORMAT, p2i(_heap_begin));
-  st->print_cr("- heap_end:                       " INTPTR_FORMAT, p2i(_heap_end));
```
Comment (calvinccheung) [inline src/hotspot/share/cds/filemap.cpp:290]: Maybe print the _heap_roots_offset?

**Review by calvinccheung — APPROVED** [review]
Just couple of nits.

File: src/hotspot/share/cds/serializeClosure.hpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  bool writing() {
+    return !reading();
+  }
+
+  // Useful alias
+  template <typename T> void do_ptr(T** p) { do_ptr((void**)p); }
+};
+
+#endif // SHARED_CDS_SERIALIZECLOSURE_HPP
```
[the remark this answers is not in the document]
Comment (iklam) [reply src/hotspot/share/cds/serializeClosure.hpp]: I fixed the `SHARED_` include guards in this file, plus a few other headers in the `share/cds/`directory. I also took this opportunity to fix some indent issues in cds_globals.hpp.

File: src/hotspot/share/cds/filemap.cpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/cds/filemap.cpp:290]: Hi Calvin, thanks for the review. I added the `_heap_roots_offset` into the debug output.

**Comment by iklam** [comment]
Thanks @ashu-mehra and @calvinccheung for the review.
/integrate

**Pull request closed without merging**